src/libs/antares/study/parallel-years.cpp

```diff
@@ -0,0 +1,431 @@
+        {
+            set->setsSizes++;
+            // Another year performed
+            ++nbYearsReallyPerformed;
+
+            // Number of actually performed years in the current set (up to now).
+            set->nbPerformedYears++;
+            // Index of the MC year's space (useful if this year is actually run)
+            indexSpace = set->nbPerformedYears - 1;
+
+            set->isYearPerformed[y] = true;
+            set->performedYearToSpace[y] = indexSpace;
+            set->spaceToPerformedYear[indexSpace] = y;
+
+            if (!foundFirstPerformedYearOfCurrentSet)
+            {
+                set->isFirstPerformedYearOfASet[y] = true;
+                foundFirstPerformedYearOfCurrentSet = true;
+            }
+        }
+        else
+        {
+            set->isYearPerformed[y] = false;
+        }
+
+        // Do we build a new set at next iteration (for years to be executed or not) ?
+
+        // In case the study is run in the draft mode, only 1 core is allowed
+        if (p.mode == Antares::Data::stdmAdequacyDraft){
+            forcedNbOfParallelYears_ = 1;
+        }
+
+        // In case parallel mode was not chosen, only 1 core is allowed
+        if (!enableParallel_ && !forceParallel_){
+            forcedNbOfParallelYears_ = 1;
+        }
+
+        if (indexSpace == forcedNbOfParallelYears_ - 1 || y == p.nbYears - 1)
+        {
+            buildNewSet = true;
+            foundFirstPerformedYearOfCurrentSet = false;
+            if (set->nbPerformedYears > forcedNbOfParallelYears_)
+                forcedNbOfParallelYears_ = set->nbPerformedYears;
+        }
+        else
+            buildNewSet = false;
+
+     // End of loop over years
+
+    }
+
+}
+
+
+bool SetsOfParallelYearCalculator::allSetsParallelYearsHaveSameSize()
+{
+    if (p.initialReservoirLevels.iniLevels == Antares::Data::irlHotStart
+        && !setsOfParallelYears.empty() && forcedNbOfParallelYears_ > 1)
+    {
+        uint currentSetSize = (uint)setsOfParallelYears[0].setsSizes;
+        return all_of(setsOfParallelYears.begin(), setsOfParallelYears.end(),
+                    [currentSetSize](const setOfParallelYears& v){ return v.setsSizes == currentSetSize; } );
+    } // End if hot start
+    return true;
+    // parameters.allSetsHaveSameSize takes this result;
+}
+
+uint SetsOfParallelYearCalculator::computeMinNbParallelYears() const
+{
+    // Now finding the smallest size among all sets.
+    uint minNbYearsInParallel = forcedNbOfParallelYears_;
+    for (uint s = 0; s < setsOfParallelYears.size(); s++)
+    {
+        uint setSize = (uint)setsOfParallelYears[s].setsSizes;
+        // Empty sets are not taken into account because, on the solver side,
+        // they will contain only skipped years
+        if (setSize && (setSize < minNbYearsInParallel))
+            minNbYearsInParallel = setSize;
+    }
+    return minNbYearsInParallel;
+}
+
+void SetsOfParallelYearCalculator::computeForcedNbYearsInParallelYearSet()
```

Comment:
About `computeForcedNbYearsInParallelYearSet()` : 
We could doubt this function is useful, but it is.
Indeed, before calling this function from the **build()** function, we build the sets of parallel years by constraining the max number of years in a set with a limit, which is **forcedNbOfParallelYears_**.
So we could think that **forcedNbOfParallelYears_** is the maximum size over all the actual built sets.
But no : if we haven't enough years in the simulation to reach the limit for the size of the first set, the maximum size of sets is less than  **forcedNbOfParallelYears_**.

But still, there is a difference of meaning between : 
- the limit we want for the size of sets
- the max size of the actual sets
I think we should store these two numbers in a different data member.